src/haz3lweb/Editors.re

```diff
@@ -39,7 +39,35 @@ let put_editor = (ed: Editor.t, eds: t): t =>
     Exercises(n, specs, Exercise.put_editor(exercise, ed))
   };
 
-let get_zipper = (editors: t): Zipper.t => get_editor(editors).state.zipper;
+let update = (f: Editor.t => Editor.t, editors: t): t =>
+  editors |> get_editor |> f |> put_editor(_, editors);
```

Comment:
Can be implemented with update_opt

Reply:
feel awkward to me. what do you do in the case where update_opt returns none? or i might be misunderstanding

Reply:
On the first readthrough I missed that the return types are different. This feels gross:
```reasonml
let map_editor = (f, editors: t) => {
  editors |> get_editor |> f
};

let update = (f: Editor.t => Editor.t, editors: t): t =>
  map_editor(f, editors) |> put_editor(_, editors)

let update_opt = (editors: t, f: Editor.t => option(Editor.t)): option(t) =>
  map_editor(f, editors) |> Option.map(put_editor(_, editors));
  ```

Reply:
What you have now feels great given the options. This problem kind of reminds me of https://reasonablypolymorphic.com/blog/higher-kinded-data/ where you want to interleave an effect through the data